components/segmenter/src/complex.rs

```diff
@@ -4,50 +4,65 @@
 
 use crate::dictionary::DictionarySegmenter;
 use crate::language::*;
+#[cfg(feature = "lstm")]
+use crate::lstm::LstmSegmenter;
 use crate::provider::*;
 use alloc::vec::Vec;
 use icu_locid::{locale, Locale};
 use icu_provider::prelude::*;
 
+#[cfg(not(feature = "lstm"))]
+type DictOrLstm = Result<DataPayload<UCharDictionaryBreakDataV1Marker>, core::convert::Infallible>;
+#[cfg(not(feature = "lstm"))]
+type DictOrLstmBorrowed<'a> =
+    Result<&'a DataPayload<UCharDictionaryBreakDataV1Marker>, &'a core::convert::Infallible>;
```

Comment:
Nit: You should almost never have a

`&'a DataPayload<FooV1Marker>`

you should instead have a

`&'a FooV1<'a>`